File: jvm/core/main/scala/com/webank/eggroll/core/deepspeed/store/meta/RendezvousStoreGetResponse.scala
package com.webank.eggroll.core.deepspeed.store.meta

import com.google.protobuf.ByteString
import com.webank.eggroll.core.meta.Deepspeed

import scala.language.implicitConversions

case class RendezvousStoreGetResponse(value: V, isTimeout: Boolean)

object RendezvousStoreGetResponse {
  implicit def serialize(src: RendezvousStoreGetResponse): Array[Byte] = {
    val builder = Deepspeed.StoreGetResponse.newBuilder()
      .setValue(ByteString.copyFrom(src.value.toArray))
      .setIsTimeout(src.isTimeout)
    builder.build().toByteArray
  }

  implicit def deserialize(byteString: ByteString): RendezvousStoreGetResponse = {
    val src = Deepspeed.StoreGetResponse.parseFrom(byteString)
    RendezvousStoreGetResponse(
      value = src.getValue.toByteArray.toVector,
      isTimeout = src.getIsTimeout
    )
  }
}

File: jvm/new_core/src/main/java/com/eggroll/core/deepspeed/store/RendezvousStoreGetResponse.java
package com.eggroll.core.deepspeed.store;

import com.eggroll.core.utils.JsonUtil;
import com.google.protobuf.ByteString;
import com.google.protobuf.InvalidProtocolBufferException;
import com.webank.eggroll.core.meta.Deepspeed;

public class RendezvousStoreGetResponse<V> {
    private V value;
    private boolean isTimeout;

    public RendezvousStoreGetResponse(V value, boolean isTimeout) {
        this.value = value;
        this.isTimeout = isTimeout;
    }

    public V getValue() {
        return value;
    }

    public void setValue(V value) {
        this.value = value;
    }

    public boolean isTimeout() {
        return isTimeout;
    }

    public void setTimeout(boolean isTimeout) {
        this.isTimeout = isTimeout;
    }

    public static <V> byte[] serialize(RendezvousStoreGetResponse<V> src) {
        Deepspeed.StoreGetResponse.Builder builder = Deepspeed.StoreGetResponse.newBuilder();
        builder.setValue(ByteString.copyFrom(JsonUtil.convertToByteArray(src.getValue())));
        builder.setIsTimeout(src.isTimeout);
        return builder.build().toByteArray();
    }

    public static <V> RendezvousStoreGetResponse<V> deserialize(byte[] bytes) throws InvalidProtocolBufferException {
        Deepspeed.StoreGetResponse src = Deepspeed.StoreGetResponse.parseFrom(bytes);
        return new RendezvousStoreGetResponse<>(
                (V) src.getValue(),
                src.getIsTimeout()
        );
    }

}